packages/discovery/src/discovery/handlers/user/StateFromEventHandler.ts

```diff
@@ -0,0 +1,94 @@
+import { ContractValue, EthereumAddress } from '@l2beat/shared'
+import { utils } from 'ethers'
+import { pick, reduce } from 'lodash'
+import * as z from 'zod'
+
+import { DiscoveryLogger } from '../../DiscoveryLogger'
+import { DiscoveryProvider } from '../../provider/DiscoveryProvider'
+import { Handler, HandlerResult } from '../Handler'
+import { getEventFragment } from '../utils/getEventFragment'
+import { toContractValue } from '../utils/toContractValue'
+
+export type StateFromEventDefinition = z.infer<typeof StateFromEventDefinition>
+export const StateFromEventDefinition = z.strictObject({
+  type: z.literal('stateFromEvent'),
+  event: z.string(),
+  returnParams: z.array(z.string()),
+  groupBy: z.optional(z.string()),
+  ignoreRelative: z.optional(z.boolean()),
+})
+
+export class StateFromEventHandler implements Handler {
+  readonly dependencies: string[] = []
+  private readonly fragment: utils.EventFragment
+  private readonly abi: utils.Interface
+
+  constructor(
+    readonly field: string,
+    readonly definition: StateFromEventDefinition,
+    abi: string[],
+    readonly logger: DiscoveryLogger,
+  ) {
+    this.fragment = getEventFragment(definition.event, abi, () => true)
+    this.abi = new utils.Interface([this.fragment])
+  }
+
+  getEvent() {
+    return this.fragment.format(utils.FormatTypes.full)
+  }
+
+  async execute(
+    provider: DiscoveryProvider,
+    address: EthereumAddress,
+    blockNumber: number,
+  ): Promise<HandlerResult> {
+    this.logger.logExecution(this.field, ['Querying ', this.fragment.name])
+    const logs = await provider.getLogs(
+      address,
+      [this.abi.getEventTopic(this.fragment)],
+      0,
+      blockNumber,
+    )
+
+    const values = new Set<ContractValue>()
+    for (const log of logs) {
+      const parsed = this.abi.parseLog(log)
+
+      const params = reduce(
+        pick(parsed.args, this.definition.returnParams),
+        (acc, value, key) => {
+          acc[key] = toContractValue(value)
+          return acc
+        },
+        {} as Record<string, ContractValue>,
+      )
+      values.add(params)
+    }
+
+    if (typeof this.definition.groupBy !== 'undefined') {
+      const result = reduce(
+        [...values],
+        (acc, value) => {
+          const groupByKey = String(
+            value[this.definition.groupBy as keyof typeof value],
```

Comment:
Let's check that on runtime. We have no guarantee that `this.definition.groupBy` is in fact `keyof typeof value`